services/listen_dnsport.c

```diff
@@ -1153,30 +1191,412 @@ make_sock_port(int stype, const char* ifname, int port,
+			coap_context, dtls_psk, dtls_psk_len, dtls_psk_hint,
+			dtls_pki_key, dtls_pki_pem, dtls_pki_verifypem) < 0) {
+		return NULL;
+	}
+	if (doc_setup_oscore(coap_context, oscore_conf, oscore_seq_file) < 0) {
+		return NULL;
+	}
+
+	return coap_context;
+}
+
+coap_bin_const_t* doc_copy_coap_bin_const_t(const coap_bin_const_t* original) {
+	if (original == NULL || original->s == NULL || original->length == 0) {
+		/* No valid data to copy */
+		return NULL;
+	}
+
+	coap_bin_const_t* copy = (coap_bin_const_t*)malloc(sizeof(coap_bin_const_t));
+	if (copy == NULL) {
+		return NULL;
+	}
+
+	uint8_t* s = (uint8_t*)malloc(original->length);
+	if (s == NULL) {
+		free(copy);
+		return NULL;
+	}
+
+	memcpy(s, original->s, original->length);
+	copy->s = s;
+	copy->length = original->length;
+
+	return copy;
+}
+
+static void
+doc_extract_pdu_info(const coap_pdu_t *pdu, struct coap_pdu_response_data *data) {
+	if (!pdu || !data) return;
+
+	data->type = coap_pdu_get_type(pdu);
+	data->code = coap_pdu_get_code(pdu);
+
+	data->mid = coap_pdu_get_mid(pdu);
+
+	coap_bin_const_t original_token = coap_pdu_get_token(pdu);
+	coap_bin_const_t* token = doc_copy_coap_bin_const_t(&original_token);
+
+	data->token = token;
+
+	coap_opt_iterator_t opt_iter;
+	coap_option_iterator_init((coap_pdu_t*)pdu, &opt_iter, COAP_OPT_ALL);
+	coap_opt_t* option;
+	while ((option = coap_option_next(&opt_iter))) {
+		coap_insert_optlist(&data->options, coap_new_optlist(opt_iter.number, coap_opt_length(option), coap_opt_value(option)));
+	}
+}
+
+static void
+doc_handle_fetch(coap_resource_t *resource, coap_session_t *session,
+		const coap_pdu_t *request, const coap_string_t *query,
+		coap_pdu_t *response) {
+
+	struct comm_point* cp = coap_resource_get_userdata(resource);
+
+	size_t size;
+	const uint8_t* data;
+	struct comm_reply rep;
+	struct coap_pdu_response_data* pdu_wrapper = (struct coap_pdu_response_data*)malloc(sizeof(struct coap_pdu_response_data));
+	rep.c = (struct comm_point*)cp;
+	rep.session = session;
+	rep.response = response;
+
+	struct sldns_buffer* buffer;
+
+	doc_extract_pdu_info(response, pdu_wrapper);
+
+	rep.pdu_wrapper = pdu_wrapper;
+
+	/* Get remote address for the CoAP session */
+	const coap_address_t* remote_addr = coap_session_get_addr_remote(session);
+
+	/* Ensure that remote IPv4 or IPv6 addresses are set */
+	if (remote_addr->addr.sa.sa_family == AF_INET) {
+		/* Copy IPv4 address */
+		memcpy(&rep.remote_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in));
+		rep.remote_addrlen = sizeof(struct sockaddr_in);
+		memcpy(&rep.client_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in));
+		rep.client_addrlen= sizeof(struct sockaddr_in);
+	} else if (remote_addr->addr.sa.sa_family == AF_INET6) {
+		/* Copy IPv6 address */
+		memcpy(&rep.remote_addr, &remote_addr->addr.sin6, sizeof(struct sockaddr_in6));
+		rep.remote_addrlen = sizeof(struct sockaddr_in6);
+		memcpy(&rep.client_addr, &remote_addr->addr.sin, sizeof(struct sockaddr_in6));
```

Comment:
Copy the IPv6 address.
```suggestion
		memcpy(&rep.client_addr, &remote_addr->addr.sin6, sizeof(struct sockaddr_in6));
```